File: week-3/pow-x-n-dp/pow-x-n-dp.js
/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function(x, n) {
    if (n < 0) {
        n = -n;
        x = 1/x;
    }
    
    let power = 1;
    while (n) {
        if (n%2 == 1) power *= x;
        x *= x;
        n = Math.floor(n/2);
    }
    
    return power;
};

File: week-3/pow-x-n-dp/pow-x-n-dp.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            n = -n
            x = 1/x
        power = 1
        while n:
            if n%2 == 1:
                power *= x
            x *= x
            n //= 2
        
        return power
